session/txn.go

```diff
@@ -270,8 +287,18 @@ func (txn *LazyTxn) changeToInvalid() {
 	txn.txnFuture = nil
 
 	txn.mu.Lock()
-	defer txn.mu.Unlock()
+	lastState := txn.mu.TxnInfo.State
+	lastStateChangeTime := txn.mu.TxnInfo.LastStateChangeTime
+	hasLock := !txn.mu.TxnInfo.BlockStartTime.IsZero()
 	txn.mu.TxnInfo = txninfo.TxnInfo{}
+	txn.mu.Unlock()
+	if !lastStateChangeTime.IsZero() {
+		hasLockLbl := "false"
+		if hasLock {
+			hasLockLbl = "true"
+		}
+		metrics.TxnDurationHistogram.WithLabelValues(txninfo.StateLabel[lastState], hasLockLbl).Observe(time.Since(lastStateChangeTime).Seconds())
```

Comment:
We can either update metrics here or when the txn is done, I don't know which one is preferred.
The benifits of update it here is it contains in progress transaction's info and consumes slightly less memory.

Reply:
Does it also make a difference in the case of retrying a transaction?